File: platform_src/android/app/src/main/java/org/elastos/trinity/dapprt/AppInstaller.java
 /*
  * Copyright (c) 2018 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

package org.elastos.trinity.dapprt;

import android.content.Context;
import android.content.res.AssetManager;
import android.net.Uri;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.Random;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class AppInstaller {

    final String[] pluginWhitelist = {
            "device",
            "networkstatus",
            "splashscreen",
    };

    final String[] urlWhitelist = {
            "http://www.elastos.org/*",
    };

    private String appPath = null;
    private String dataPath = null;
    private ManagerDBAdapter dbAdapter = null;
    private Context context = null;

    private Random random =new Random();

    public boolean init(Context context, ManagerDBAdapter dbAdapter, String appPath, String dataPath) {
        this.context = context;
        this.appPath = appPath;
        this.dataPath = dataPath;
        this.dbAdapter = dbAdapter;

        random = new Random();

        return true;
    }

    private boolean unpackZip(InputStream srcZip, String destPath) {
        InputStream is;
        ZipInputStream zis;

        try
        {
            String filepath;
            zis = new ZipInputStream(new BufferedInputStream(srcZip));
            ZipEntry ze;
            byte[] buffer = new byte[1024];
            int count;

            while ((ze = zis.getNextEntry()) != null) {
                filepath = destPath + ze.getName();

                if (ze.isDirectory()) {
                    File fmd = new File(filepath);
                    fmd.mkdirs();
                }
                else {
                    File file = new File(filepath);
                    file.getParentFile().mkdirs();

                    FileOutputStream fout = new FileOutputStream(file);

                    // cteni zipu a zapis
                    while ((count = zis.read(buffer)) != -1) {
                        fout.write(buffer, 0, count);
                    }

                    fout.close();
                }
                zis.closeEntry();
            }

            zis.close();
        }
        catch(IOException e) {
            e.printStackTrace();
            return false;
        }

        return true;
    }

    private boolean downloadDAppPackage(String url, String destFile) throws Exception {
        BufferedInputStream in = new BufferedInputStream(new URL(url).openStream());
        FileOutputStream fileOutputStream = new FileOutputStream(destFile);
        byte dataBuffer[] = new byte[1024];
        int bytesRead;
        while ((bytesRead = in.read(dataBuffer, 0, 1024)) != -1) {
            fileOutputStream.write(dataBuffer, 0, bytesRead);
        }
        return true;
    }

    private void deleteDAppPackage(String packagePath) {
        if (packagePath != null && !packagePath.isEmpty()) {
            File file = new File(packagePath);
            file.delete();
        }
    }

    public AppInfo install(AppManager appManager, String url) throws Exception {
        InputStream inputStream = null;
        AppInfo info = null;
        String downloadPkgPath = null;

        if (url.startsWith("assets://")) {
            AssetManager manager = context.getAssets();
            String substr = url.substring(9);
            inputStream = manager.open(substr);
        }
        else if (url.startsWith("content://")) {
            Uri uri = Uri.parse(url);
            inputStream = context.getContentResolver().openInputStream(uri);
        }
        else if (url.startsWith("http://") || url.startsWith("https://")) {
            downloadPkgPath = appPath + "tmp_" + random.nextInt() + ".epk";
            if (downloadDAppPackage(url, downloadPkgPath)) {
                inputStream = new FileInputStream(downloadPkgPath);
            }
        }
        else {
            if (url.startsWith("file://")) {
                url = url.substring(7);
            }
            inputStream = new FileInputStream(url);
        }

        String temp = "tmp_" + random.nextInt();
        String path = appPath + temp + "/";

        File fmd = new File(path);
        fmd.mkdirs();

        if (!unpackZip(inputStream, path)) {
            deleteDAppPackage(downloadPkgPath);
            throw new Exception("UnpackZip fail!");
        }

        String manifest = path + "manifest.json";
        InputStream file = null;
        file = new FileInputStream(manifest);

        info = parseManifest(file);
        File from = new File(appPath, temp);
        if (info == null || info.app_id == null || info.app_id.equals("launcher")
                || appManager.getAppInfo(info.app_id) != null) {
            deleteAllFiles(from);
            deleteDAppPackage(downloadPkgPath);
            throw new Exception("App alreadey exist!");
        }
        else {
            File to = new File(appPath, info.app_id);
            if (to.exists()) {
                deleteAllFiles(to);
                to = new File(appPath, info.app_id);
            }
            from.renameTo(to);
            info.built_in = 0;
            dbAdapter.addAppInfo(info);
            deleteDAppPackage(downloadPkgPath);
            return info;
        }
    }

    private boolean deleteAllFiles(File root) throws Exception {
        if (!root.exists()) {
            return false;
        }

        File files[] = root.listFiles();
        if (files != null) {
            for (File f : files) {
                if (f.isFile() && f.exists()) {
                    f.delete();
                }
                else {
                    deleteAllFiles(f);
                }
            }
        }

        root.delete();

        return true;
    }

    public void unInstall(AppInfo info)  throws Exception {
        if (info == null) {
            throw new Exception("No such app!");
        }

        if (info.built_in == 1) {
            throw new Exception("App is a built in!");
        }
        int count = dbAdapter.removeAppInfo(info);
        if (count < 1) {
            throw new Exception("Databashe error!");
        }

        File root = new File(appPath + info.app_id);
        deleteAllFiles(root);
        root = new File(dataPath + info.app_id);
        deleteAllFiles(root);
    }

    private boolean isAllowPlugin(String name) {
        String pluginName = name.toLowerCase();
        for (String item : pluginWhitelist) {
            if (item.equals(pluginName)) {
                return true;
            }
        }
        return false;
    }

    private boolean isAllowUrl(String url) {
        url = url.toLowerCase();
        for (String item : urlWhitelist) {
            if (item.equals(url)) {
                return true;
            }
        }
        return false;
    }


    private String getMustStrValue(JSONObject json, String name) throws Exception {
        if (json.has(name)) {
            return json.getString(name);
        } else {
            throw new Exception("Parse Manifest.json error: '" + name + "' no exist!");
        }
    }

    public AppInfo parseManifest(InputStream inputStream) throws Exception {
        AppInfo appInfo = new AppInfo();

        InputStreamReader inputStreamReader = new InputStreamReader(inputStream, "UTF-8");
        BufferedReader bufReader = new BufferedReader(inputStreamReader);
        String line;
        StringBuilder builder = new StringBuilder();
        while ((line = bufReader.readLine()) != null) {
            builder.append(line);
        }
        bufReader.close();
        inputStreamReader.close();


        JSONObject json = new JSONObject(builder.toString());

        //Must
        appInfo.app_id = getMustStrValue(json, "id");
        appInfo.version = getMustStrValue(json, "version");
        appInfo.name = getMustStrValue(json, "name");
        appInfo.start_url = getMustStrValue(json, "start_url");

        if (json.has("icons")) {
            JSONArray array = json.getJSONArray("icons");
            for (int i = 0; i < array.length(); i++) {
                JSONObject icon = array.getJSONObject(i);
                String src = icon.getString(AppInfo.SRC);
                String sizes = icon.getString(AppInfo.SIZES);
                String type = icon.getString(AppInfo.TYPE);
                appInfo.addIcon(src, sizes, type);
            }
        }
        else {
            throw new Exception("Parse Manifest.json error: 'icons' no exist!");
        }

        //Optional
        if (json.has(AppInfo.SHORT_NAME)) {
            appInfo.short_name = json.getString(AppInfo.SHORT_NAME);
        }

        if (json.has(AppInfo.DESCRIPTION)) {
            appInfo.description = json.getString(AppInfo.DESCRIPTION);
        }

        if (json.has(AppInfo.DEFAULT_LOCAL)) {
            appInfo.default_locale = json.getString(AppInfo.DEFAULT_LOCAL);
        }

        if (json.has("author")) {
            JSONObject author = json.getJSONObject("author");
            if (author.has("name")) {
                appInfo.author_name = author.getString("name");
            }
            if (author.has("email")) {
                appInfo.author_email = author.getString("email");
            }
        }

        if (json.has("plugins")) {
            JSONArray array = json.getJSONArray("plugins");
            for (int i = 0; i < array.length(); i++) {
                String plugin = array.getString(i);
                int authority = AppInfo.AUTHORITY_NOINIT;
                if (isAllowPlugin(plugin)) {
                    authority = AppInfo.AUTHORITY_ALLOW;
                }
                appInfo.addPlugin(plugin, authority);
            }
        }

        if (json.has("urls")) {
            JSONArray array = json.getJSONArray("urls");
            for (int i = 0; i < array.length(); i++) {
                String url = array.getString(i);
                int authority = AppInfo.AUTHORITY_NOINIT;
                if (isAllowUrl(url)) {
                    authority = AppInfo.AUTHORITY_ALLOW;
                }
                appInfo.addUrl(url, authority);
            }
        }

        if (json.has(AppInfo.BACKGROUND_COLOR)) {
            appInfo.background_color = json.getString(AppInfo.BACKGROUND_COLOR);
        }

        if (json.has("theme")) {
            JSONObject theme = json.getJSONObject("theme");
            if (theme.has(AppInfo.THEME_DISPLAY)) {
                appInfo.theme_display = theme.getString(AppInfo.THEME_DISPLAY);
            }
            if (theme.has(AppInfo.THEME_COLOR)) {
                appInfo.theme_color = theme.getString(AppInfo.THEME_COLOR);
            }
            if (theme.has(AppInfo.THEME_FONT_NAME)) {
                appInfo.theme_font_name = theme.getString(AppInfo.THEME_FONT_NAME);
            }
            if (theme.has(AppInfo.THEME_FONT_COLOR)) {
                appInfo.theme_font_color = theme.getString(AppInfo.THEME_FONT_COLOR);
            }
        }

        appInfo.install_time = System.currentTimeMillis() / 1000;

        return appInfo;
    }
}

File: platform_src/ios/Elastos/Classes/AppInstaller.swift
 /*
  * Copyright (c) 2018 Elastos Foundation
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
  * of this software and associated documentation files (the "Software"), to deal
  * in the Software without restriction, including without limitation the rights
  * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  * copies of the Software, and to permit persons to whom the Software is
  * furnished to do so, subject to the following conditions:
  *
  * The above copyright notice and this permission notice shall be included in all
  * copies or substantial portions of the Software.
  *
  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */

 import Foundation

 extension Data {
    private static let hexAlphabet = "0123456789abcdef".unicodeScalars.map { $0 }

    public func hexEncodedString() -> String {
        return String(self.reduce(into: "".unicodeScalars, { (result, value) in
            result.append(Data.hexAlphabet[Int(value/16)])
            result.append(Data.hexAlphabet[Int(value%16)])
        }))
    }
 }

 class AppInstaller {

    let pluginWhitelist = [
        "device",
        "networkstatus",
        "splashscreen",
        ];

    let urlWhitelist = [
        "http://www.elastos.org/*",
        ];

    var appPath: String = "";
    var dataPath: String = "";
    var tempPath: String = "";
    var dbAdapter: ManagerDBAdapter;

    init(_ appPath: String, _ dataPath: String, _ tempPath: String, _ dbAdapter: ManagerDBAdapter) {
        self.appPath = appPath;
        self.dataPath = dataPath;
        self.tempPath = tempPath;
        self.dbAdapter = dbAdapter;
    }

    func unpackZip(_ srcZip: String, _ destPath: String) -> Bool {
        return SSZipArchive.unzipFile(atPath: srcZip, toDestination: destPath);
    }

    private func sha256(data : Data) -> Data {
        var hash = [UInt8](repeating: 0,  count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes {
            _ = CC_SHA256($0, CC_LONG(data.count), &hash)
        }
        return Data(bytes: hash)
    }

    private func verifyEpkDigest(_ destPath: String) -> Bool {
        let fileManager = FileManager.default

        let dirEnum = fileManager.enumerator(atPath: destPath)
        var fileHashMap = [String: String]()
        var fileListSHA = "EPK-SIGN/FILELIST.SHA not found"
        while let filePath = dirEnum?.nextObject() as? String {
            var isDir : ObjCBool = false
            let fileURL =  NSURL(fileURLWithPath: destPath).appendingPathComponent(filePath)!
            if (fileManager.fileExists(atPath: fileURL.path, isDirectory: &isDir) && !isDir.boolValue) {
                do {
                    let fileData = try Data(contentsOf: fileURL)
                    let hash = sha256(data: fileData)
                    let hashString = hash.hexEncodedString()
                    if (!filePath.starts(with: "EPK-SIGN/")) {
                        fileHashMap[filePath] = hashString
                    } else if (filePath == "EPK-SIGN/FILELIST.SHA") {
                        fileListSHA = String(data: fileData, encoding: .utf8)!
                    }
                }
                catch {
                    print("Failed to hash file " + fileURL.path)
                    return false
                }
            }
        }

        var fileList = ""
        for (k,v) in fileHashMap.sorted(by: {$0.0 < $1.0}) {
            fileList = "\(fileList)\(v) \(k)\n"
        }

        let epkHash = sha256(data: fileList.data(using: .utf8)!).hexEncodedString()
        if (epkHash == fileListSHA) {
            print("Matched EPK digest with \(fileListSHA)")
            return true
        }
        print("Failed to match EPK digest")
        return false;
    }

    private func contentsOf(file fileURL: NSURL) -> String? {
        let fileManager = FileManager.default

        var isDir : ObjCBool = false
        if (fileManager.fileExists(atPath: fileURL.path!, isDirectory: &isDir) && !isDir.boolValue) {
            do {
                let fileData = try Data(contentsOf: fileURL as URL)
                return String(data: fileData, encoding: .utf8)
            }
            catch {
                print("Failed to read file " + fileURL.path!)
                return nil
            }
        }

        print("Failed to read file " + fileURL.path!)
        return nil
    }

    private func verifyEpkSignature(_ destPath: String) -> Bool {
        let publicKey = contentsOf(file: NSURL(fileURLWithPath: destPath).appendingPathComponent("EPK-SIGN/SIGN.PUB")! as NSURL)
//        let payload = contentsOf(file: NSURL(fileURLWithPath: destPath).appendingPathComponent("EPK-SIGN/FILELIST.SHA")! as NSURL)
//        let signed_payload = contentsOf(file: NSURL(fileURLWithPath: destPath).appendingPathComponent("EPK-SIGN/FILELIST.SIGN")! as NSURL)

        //TODO:: need to check
//        let succeeded = ela_verify_message(publicKey, payload, signed_payload)
        let succeeded = true;

        if (succeeded) {
            print("Successfully verified EPK signature")
            print("The public key of the EPK is \(publicKey!)")
        } else {
            print("Failed to verify EPK signature")
        }
        return succeeded
    }

    func deleteAllFiles(_ path: String) throws {
        let fileManager = FileManager.default;
        try fileManager.removeItem(atPath: path)
    }

    func copyAssetsFolder(_ src: String, _ dest: String) throws {
        let fileManager = FileManager.default
        if fileManager.fileExists(atPath: dest){
            try fileManager.removeItem(atPath: dest)
        }

        try fileManager.copyItem(atPath: src, toPath: dest)
    }

    func  install(_ appManager: AppManager, _ url: String) throws -> AppInfo? {
        var zipPath = url;
        if (url.hasPrefix("asset://")) {
            zipPath = getAssetPath(url);
        }
        else if (url.hasPrefix("file://")) {
            let index = url.index(url.startIndex, offsetBy: 7)
            zipPath = String(url[index ..< url.endIndex]);
        }

        let temp = "tmp_" + UUID().uuidString
        let temPath = appPath + temp;

        if (!unpackZip(zipPath, temPath)) {
            throw AppError.error("UnpackZip fail!");
        }

        if (!verifyEpkDigest(temPath)) {
            throw AppError.error("verifyEpkDigest fail!");
        }

        if (!verifyEpkSignature(temPath)) {
            throw AppError.error("verifyEpkSignature fail!");
        }

        let fileManager = FileManager.default;

        let info = try getInfoByManifest(temPath);
        guard (info != nil && info!.app_id != "" && info!.app_id != "launcher") else {
            try deleteAllFiles(temPath);
            throw AppError.error("App info error!");
        }
        
        guard (appManager.getAppInfo(info!.app_id) == nil) else {
            try deleteAllFiles(temPath);
            throw AppError.error("App '" + info!.app_id + "' alreadey exist!");
        }

        let path = appPath + info!.app_id;
        if (fileManager.fileExists(atPath: path)) {
            try deleteAllFiles(path);
        }

        try fileManager.moveItem(atPath: temPath, toPath: path);
//            let dirs = try! fileManager.contentsOfDirectory(atPath: path);

        info!.built_in = false;
        try appManager.dbAdapter.addAppInfo(info!);
        return info!;
    }

    func unInstall(_ info: AppInfo?) throws {
        guard info != nil else {
            throw AppError.error("No such app!");
        }

        guard !info!.built_in else {
            throw AppError.error("App is a built in!");
        }

        let dataPath = self.dataPath + info!.app_id
        try deleteAllFiles(dataPath);
        try dbAdapter.removeAppInfo(info!);
        let appPath = self.appPath + info!.app_id
        try deleteAllFiles(appPath);
    }

    private func isAllowPlugin(_ plugin: String) -> Bool {
        for item in pluginWhitelist {
            if (item == plugin) {
                return true;
            }
        }
        return false;
    }

    private func isAllowUrl(_ url: String) -> Bool {
        for item in urlWhitelist {
            if (item == url) {
                return true;
            }
        }
        return false;
    }

    func getInfoByManifest(_ path: String) throws -> AppInfo? {
        var dir = path;
        let fileManager = FileManager.default;
        var ret = fileManager.fileExists(atPath: dir + "/manifest.json")
        if (!ret) {
            dir = dir + "/assets";
            ret = fileManager.fileExists(atPath: dir + "/manifest.json");
            guard ret else {
                try deleteAllFiles(path);
                throw AppError.error("manifest.json no exist!");
            }
        }
        let info = try parseManifest(dir + "/manifest.json");
        ret = fileManager.fileExists(atPath: dir + "/manifest.i18n");
        if (ret) {
            try parseManifestLocale(dir + "/manifest.i18n", info!);
        }
        return info;
    }

    private func getMustStrValue(_ json: [String: Any], _ name: String) throws -> String {
        let value = json[name] as? String;
        if (value != nil) {
            return value!
        }
        else {
            throw AppError.error("Parse Manifest.json error: '\(name)' no exist!");
        }
    }

    private func getMustIntValue(_ json: [String: Any], _ name: String) throws -> Int {
        let value = json[name] as? Int;
        if (value != nil) {
            return value!
        }
        else {
            throw AppError.error("Parse Manifest.json error: '\(name)' no exist!");
        }
    }
    
    func parseManifest(_ path: String, _ launcher:Bool = false) throws -> AppInfo? {
        let appInfo = AppInfo();
        let url = URL.init(fileURLWithPath: path)
        var value: String?;

        let data = try Data(contentsOf: url);
        let json = try JSONSerialization.jsonObject(with: data,
                                                    options: []) as! [String: Any];

        //Must
        appInfo.app_id = try getMustStrValue(json, "id");
        appInfo.version = try getMustStrValue(json, "version");
        appInfo.version_code = try getMustIntValue(json, "version_code");
        appInfo.name = try getMustStrValue(json, "name");
        appInfo.start_url = try getMustStrValue(json, "start_url");
        let range = appInfo.start_url.range(of: "://");
        if range != nil{
            appInfo.remote = true;
        }
        else {
            appInfo.remote = false;
        }

        let icons = json["icons"] as? [Dictionary<String, String>];
        if !launcher {
            if icons != nil {
                for icon in icons! {
                    let src = icon["src"];
                    let sizes = icon["sizes"];
                    let type = icon["type"];
                    appInfo.addIcon(src!, sizes!, type!);
                }
            }
            else {
                throw AppError.error("Parse Manifest.json error: 'icons' no exist!");
            }
        }

        //Optional
        value = json[AppInfo.START_VISIBLE] as? String;
        if value != nil {
            appInfo.start_visible = value!;
        }
        else {
            appInfo.start_visible = "show";
        }
        
        value = json["short_name"] as? String;  
        if value != nil {
            appInfo.short_name = value!;
        }

        value = json["description"] as? String;
        if value != nil {
            appInfo.desc = value!;
        }

        value = json["default_locale"] as? String;
        if value != nil {
            appInfo.default_locale = value!;
        }
        else {
            appInfo.default_locale = "en";
        }

        value = json["type"] as? String;
        if value != nil {
            appInfo.type = value!;
        }
        else {
            appInfo.type = "app";
        }

        let author = json["author"] as? [String: Any];
        if author != nil {
            value = author!["name"] as? String;
            if value != nil {
                appInfo.author_name = value!;
            }
            value = author!["email"] as? String;
            if value != nil {
                appInfo.author_email = value!;
            }
        }
        
        value = json["category"] as? String;
        if value != nil {
            appInfo.category = value!;
        }
        else {
            appInfo.category = "other";
        }
        
        value = json["key_words"] as? String;
        if value != nil {
            appInfo.key_words = value!;
        }
        else {
            appInfo.key_words = "";
        }

        var authority = AppInfo.AUTHORITY_NOINIT;
        let plugins = json["plugins"] as? [String];
        if (plugins != nil) {
            for plugin in plugins! {
                authority = AppInfo.AUTHORITY_NOINIT;
                let pluginName = plugin.lowercased();
                if (isAllowPlugin(pluginName)) {
                    authority = AppInfo.AUTHORITY_ALLOW;
                }
                appInfo.addPlugin(pluginName, authority);
            }
        }

        let urls = json["urls"] as? [String];
        if (urls != nil) {
            for url in urls! {
                authority = AppInfo.AUTHORITY_NOINIT;
                let urlString = url.lowercased();
                if (isAllowUrl(urlString)) {
                    authority = AppInfo.AUTHORITY_ALLOW;
                }
                appInfo.addUrl(urlString, authority);
            }
        }

        let intents = json["intents"] as? [String];
        if (intents != nil) {
            for intent in intents! {
                authority = AppInfo.AUTHORITY_ALLOW;
                let urlString = intent.lowercased();
                appInfo.addIntent(urlString, authority);
            }
        }
        
        let frameworks = json["framework"] as? [String];
        if (frameworks != nil) {
            for framework in frameworks! {
                let element = framework.components(separatedBy: "@");
                if (element.count == 1) {
                    appInfo.addFramework(element[0], "");
                }
                else if (element.count > 1) {
                    appInfo.addFramework(element[0], element[1]);
                }
            }
        }

        let platforms = json["platform"] as? [String];
        if (platforms != nil) {
            for platform in platforms! {
                let element = platform.components(separatedBy: "@");
                if (element.count == 1) {
                    appInfo.addPlatform(element[0], "");
                }
                else if (element.count > 1) {
                    appInfo.addPlatform(element[0], element[1]);
                }
            }
        }


        value = json["background_color"] as? String;
        if value != nil {
            appInfo.background_color = value!;
        }

        let theme = json["theme"] as? [String: Any];
        if (theme != nil) {
            value = theme!["display"] as? String;
            if value != nil {
                appInfo.theme_display = value!;
            }
            value = theme!["color"] as? String;
            if value != nil {
                appInfo.theme_color = value!;
            }

            value = theme!["font_name"] as? String;
            if value != nil {
                appInfo.theme_font_name = value!;
            }

            value = theme!["font_color"] as? String;
            if value != nil {
                appInfo.theme_font_color = value!;
            }
        }

        let intent_filters = json["intent_filters"] as? [Dictionary<String, String>];
        if intents != nil {
            for intent in intent_filters! {
                let action = intent["action"];
                if (action != nil) {
                    appInfo.addIntentFilter(action!);
                }
            }
        }

        appInfo.install_time = Int64(Date().timeIntervalSince1970);
        appInfo.launcher = launcher;

        let fileManager = FileManager.default
        if (!fileManager.fileExists(atPath: dataPath + appInfo.app_id)) {
            do {
                try fileManager.createDirectory(atPath: dataPath + appInfo.app_id, withIntermediateDirectories: true, attributes: nil)
            }
            catch let error {
                print("Make dataPath error: \(error)");
            }
        }
        if (!fileManager.fileExists(atPath: tempPath + appInfo.app_id)) {
            do {
                try fileManager.createDirectory(atPath: tempPath + appInfo.app_id, withIntermediateDirectories: true, attributes: nil)
            }
            catch let error {
                print("Make tempPath error: \(error)");
            }
        }


        return appInfo;
    }

    func parseManifestLocale(_ path: String, _ info:AppInfo) throws {
        let url = URL.init(fileURLWithPath: path)
        let data = try Data(contentsOf: url);
        let locales = try JSONSerialization.jsonObject(with: data,
                                                    options: []) as! [String: Any];

        var exist:Bool = false;
        for (lang, dict) in locales {
            let locale = dict as! [String: String];
            let name = try getMustStrValue(locale, "name");
            let short_name = try getMustStrValue(locale, "short_name");
            let description = try getMustStrValue(locale, "author_name");
            let author_name = try getMustStrValue(locale, "author_name");
            info.addLocale(lang, name, short_name, description, author_name);

            if (lang == info.default_locale) {
                exist = true;
            }
        }

        if (!exist) {
            info.addLocale(info.default_locale, info.name, info.short_name, info.desc, info.author_name);
        }
    }
 }

